### PR TITLE
Contacts & New Contacts Page

### LINKED ISSUE #8: Create all contacts-FE
FE work
Logged in users should be able to add new contacts which are only visible to the logged in user (we don't want to expose a person's contacts to other people)

Fields:

*First Name
*Last Name
Company (dropdown that links to existing company in database)
Email
Phone Number
Notes
* Indicates required

Image

### PR DESCRIPTION
**Pull request by reneemes**
## Type of Change
- [x] feature ⛲
- [x] styling 🎨
- [x] testing 🧪

## Description

This PR adds a user contacts page that displays all contacts associated with a user in a table. Key features include a search bar to filter contacts by first name, last name, or company name. Additionally, a button directs users to an "Add New Contact" page, where they can fill out a form to create new contacts. The form includes a dropdown menu for selecting a company, which links the company name through a foreign key in the contacts table. All functionality is tested in Cypress, and all tests are passing.

## Motivation and Context

This feature enhances the user experience by providing a centralized page for managing contacts. The search functionality makes it easier for users to locate specific contacts efficiently, while the "Add New Contact" form streamlines the process of adding and associating new contacts with companies. The dropdown menu ensures data consistency by linking contacts to companies through a foreign key. Overall, this addition improves usability, data organization, and scalability for managing user contacts.

## Related Tickets

Closes #8 & 3
## Screenshots (if appropriate):

Screenshot 2024-12-19 at 4 49 18 PM
Screenshot 2024-12-19 at 4 49 27 PM

## Added Test?
- [x] Yes 🫡
  - FE End-to-end
- [ ] No 🙅
- [x] All previous tests still pass 🥳

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Review by Sgalvin36 — CHANGES_REQUESTED** [review]
All the fetch requests are currently hardcoded to go to user 4 in the components that you created. 
The rest of the code looks fantastic, well tested with cypress, and functionally separated out to make it easy to follow and understand!

File: src/components/contacts/NewContact.tsx [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+    companyId: null,
+    notes: "",
+  });
+
+  const [feedback, setFeedback] = useState<string | null>(null);
+  const [companies, setCompanies] = useState<{ id: number; name: string }[]>(
+    []
+  );
+
+  useEffect(() => {
```
Comment (jimmacur) [inline src/components/contacts/NewContact.tsx:33]: Pull this out the the trackerApiCall.tsx

File: src/components/companies/Companies.tsx [hunk 2]
```diff
@@ -29,7 +29,7 @@ function Companies() {
       try {
         const token =
           "";
-        const response = await fetch("http://localhost:3001/api/v1/users/2/companies", {
+        const response = await fetch("http://localhost:3001/api/v1/users/4/companies", {
```
[the remark this answers is not in the document]
Comment (CCirbo) [reply src/components/companies/Companies.tsx:32]: To true, great catch, thank you, we will fix now.

File: src/components/companies/NewCompany.tsx [hunk 3]
```diff
@@ -27,7 +27,7 @@ function NewCompany() {
     try {
       const token =
         "";
-      const response = await fetch("http://localhost:3001/api/v1/users/2/companies", {
+      const response = await fetch("http://localhost:3001/api/v1/users/4/companies", {
```
[the remark this answers is not in the document]
Comment (CCirbo) [reply src/components/companies/NewCompany.tsx:30]: Definitely not, thanks again, will fix!

File: src/components/layout/MenuBar.tsx
Comment (jimmacur) [inline src/components/layout/MenuBar.tsx:1]: great changes here with the order and hover functionality

**Review by jimmacur — CHANGES_REQUESTED** [review]
Things are looking amazing. And you have already dynamically connected contacts to existing companies! Nice!

Just a few changes requested.

File: src/components/contacts/NewContact.tsx [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+        body: JSON.stringify(newContact),
+      });
+
+      if (!response.ok) {
+        const errorData = await response.json();
+        throw new Error(errorData.error || "Failed to add the contact");
+      }
+      setFeedback("Contact added successfully! Redirecting...");
+      setTimeout(() => navigate("/contacts"), 3000);
+
```
[the remark this answers is not in the document]
Comment (CCirbo) [reply src/components/contacts/NewContact.tsx:108]: Yes, that is redundant and should be removed. It was, but we had some merge issues and there it is again!  Will take it out, thanks!

**Comment by jimmacur** [comment]
One more thing to consider – this can definitely be a refactor for the future – but you could enhance the phone number input by automatically formatting it with dashes as the user types.

File: src/components/contacts/NewContact.tsx (hunk 1 above)
Comment (CCirbo) [reply src/components/contacts/NewContact.tsx:33]: Great call out Jim, we have a ticket up, #62 as a refactor.

File: src/components/contacts/Contacts.tsx [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+
+function Contacts() {
+  const [contacts, setContacts] = useState<ContactData[] | []>([]);
+  const [allContacts, setAllContacts] = useState<ContactData[] | []>([]);
+  const [contactSearch, setContactSearch] = useState<string>("");
+  const [fetchError, setFetchError] = useState<string | null>(null);
+
+  const { token } = useUserLoggedContext();
+  
+  useEffect(() => {
```
[the remark this answers is not in the document]
Comment (CCirbo) [reply src/components/contacts/Contacts.tsx:36]: Great call out Jim, we have a ticket up, #62 as a refactor.

File: src/components/companies/Companies.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (reneemes) [reply src/components/companies/Companies.tsx:32]: The necessary changes have been made to include dynamic user IDs. All tests are passing as well.

File: src/components/companies/NewCompany.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (CCirbo) [reply src/components/companies/NewCompany.tsx:30]: We have fixed Contacts and New Contacts. Noticed that Companies, both pages did the same thing so I have a ticket up, #63 for their refactor.

File: src/components/contacts/Contacts.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (CCirbo) [reply src/components/contacts/Contacts.tsx:42]: Companies need to have their fetch calls refactored and moved to trackerApiCall as well, so created a ticket, #63 for that refactor and to refactor user id from hardcoded to dynamic.

**Review by jimmacur — APPROVED** [review]

**Review by Sgalvin36 — APPROVED** [review]
Thanks for the updates, everything looks good on my end now.

**Review by Sgalvin36 — CHANGES_REQUESTED** [review]
The companies fetch call is still hardcoded.

**Review by Sgalvin36 — APPROVED** [review]

**Pull request merged**